Add successful step functions

rusandris opened the pull request:
I've added the `successful_step` function to DynamicalSystemsBase discussed [here](https://github.com/JuliaDynamics/ChaosTools.jl/pull/262).

awage commented:
Hi! This is an interesting functionality, but I think this function should rely on `check_error`: there are different types of integrators that may fail with this function. 

The function `check_error` can be dispatched for the correct integrator. @Datseris what is your point of view?

rusandris commented:
This was our point: `check_error` fails for integrators that do not have the `sol.retcode` fields (like simple integrators). This is an attempt to implement a function for both simple and regular integrator types. What other types do you think this should be extended to?

awage commented:
There are special integrators such as `PoincareMap`, `StroboscopicMap` and `ProjectedIntegrator` defined in `DynamicalSystemsBase`. I don't know if they make sense for the calculation of Lyapunov exponents which is your aim, but the function should work for all integrators.

Datseris commented:
Wait, why aren't we using `check_error` here? The generic version of the function should just return `check_error` if the integrator comes from DiffEq. I agree that this function should work for all integrators we have. One way to do it is to write the general version of `successful_step(integ)` that checks if any of the state values are `Inf` or `NaN` if the integrator isn't the one from OrdinaryDiffEq.

Datseris commented:
YOu also need to add tests here. Write tests in the `test` folder that ensure that what we are doing here actually works, by initializing a dynamical system we know is unstable and checking that `successful_step` indeed returns false.

Datseris commented:
@awage were you thinking of extending `check_error` for all of our integrators? I am not sure whether this is worth the effort. How would you check for successful step in a discrete integrator besides the method I outlined above? If you have better alternatives I'm very happy to use them!

Datseris commented:
@awage were you thinking of extending `check_error` for all of our integrators? I am not sure whether this is worth the effort. How would you check for successful step in a discrete integrator besides the method I outlined above? If you have better alternatives I'm very happy to use them!

awage commented:
For projected integrators and stroboscopic maps it is easy: 
```julia
function SciMLBase.check_error(pinteg::ProjectedIntegrator) 
    return SciMLBase.check_error(pinteg.integ)
end

function SciMLBase.check_error(smap::StroboscopicMap)
    return SciMLBase.check_error(smap.integ)
end
```
For Poincaré maps it is the same. For discrete integrators I have no idea.

This is off topic but related to this function: 

I have encountered sometimes the problem when computing basins that the maximum iterations has been reached (`maxiter` keyword in `diffeq`). The problem is that the algorithm doesn't care and finds a new attractor that is an artefact. 

I wonder if a simple `check_error` inserted in the function `get_label_ic` would solve this issue. When the maximum iteration has been reached we can assign -1 or a special value to the initial condition. The algorithm can also reach the maxiter when it is in the `:att_found` state. In this case a different action can be made.

Datseris commented:
ok for the extension to projected integrators, and stroboscopic and poincare maps.

For the discrete system integrator we can use my suggestion of checking for Inf or NaN.

For our attractor finding algorithm:I guess your suggestion is the most sensible thing to do and assign -1, but could be discussed further. So let's please have this discussion in a different PR to keep this one focused. Anywyas a different PR needs to be done to alter `AttrqactorsViaRecurrences`.

rusandris commented:
>  Wait, why aren't we using `check_error` here? The generic version of the function should just return `check_error` if the integrator comes from DiffEq. 

I think `check_error` is called inside of `step!` somewhere, so if we used `check_error` again after, it would raise the warning twice. So the workaround is checking the retcodes.

Datseris commented:
> I think check_error is called inside of step! somewhere

Okay, but can you make this claim certain? Is `check_error` public API by the way? @awage where did you find `check_error`?

awage commented:
It is documented at least: 
[https://diffeq.sciml.ai/stable/basics/integrator/#SciMLBase.check_error]

rusandris commented:
> Okay, but can you make this claim certain? 

From `step!(integrator::OrdinaryDiffEq.ODEIntegrator)` in OrdinaryDiffEq /src/iterator_interface.jl

```
@inline function step!(integrator::ODEIntegrator)
  if integrator.opts.advance_to_tstop
    @inbounds while integrator.tdir * integrator.t < first(integrator.opts.tstops)
      loopheader!(integrator)
      check_error!(integrator) != :Success && return
      perform_step!(integrator,integrator.cache)
      loopfooter!(integrator)
    end
  else
    @inbounds loopheader!(integrator)
    check_error!(integrator) != :Success && return
    @inbounds perform_step!(integrator,integrator.cache)
    @inbounds loopfooter!(integrator)
    @inbounds while !integrator.accept_step
      loopheader!(integrator)
      perform_step!(integrator,integrator.cache)
      loopfooter!(integrator)
    end
  end
  @inbounds handle_tstop!(integrator)
end
```

So if you call `step!` on an unstable system using an integrator that works with `check_error`, the warnings from `check_error` show up :
```
step!(integ_from_prob,13)
┌ Warning: Instability detected. Aborting
└ @ SciMLBase ~/Documents/SciMLBase.jl/src/integrator_interface.jl:434

```

Datseris commented:
> It is documented at least: [check_error](https://diffeq.sciml.ai/stable/basics/integrator/#SciMLBase.check_error)

Okay, but @awage why did you recommend that our code should use `check_error` instead of what it does now? Furthermore, as far as the interface I'd like to have, is a function that returns booleans. `successful_step(integ) -> true/false` is what is most convenient for us.

@rusandris as far as I can tell your design is fine now, but you still need to write a generic implementation of the gfunction that always returns true if it doesn't know its input, and also add tests. Can you please do this?

awage commented:
`successful_step` is  fine, I though it was code redundancy as we could use the existing interface of `SciML`. My main point was to extend the function to all kind of integrator. I think this is important.

rusandris commented:
> @rusandris as far as I can tell your design is fine now, but you still need to write a generic implementation of the gfunction that always returns true if it doesn't know its input, and also add tests. Can you please do this?

I've added a dummy function so if the integrator is not `AbstractODEIntegrator` (this includes simple integrators, but they are handled separately) it always returns true. So there are 3 methods to `successful_step` now.

Datseris commented on src/core/continuous.jl:
```diff
@@ -283,3 +283,10 @@ function SciMLBase.reinit!(integ::AbstractODEIntegrator, u0::AbstractVector,
     reinit!(integ, integ.u; kwargs...)
 end
+
+successful_step(integ) = true
+successful_step(integ::AbstractODEIntegrator{Alg, IIP, S, T}) where {Alg <:AbstractSimpleDiffEqODEAlgorithm, IIP, S, T} = true
```
These shouldn't check for `true`. They should check if `get_state(integ)` contains `Inf` or `NaN`.

rusandris replied:
I'm afraid that wouldn't work. `step!` with the  simple integrator errors `dt<dtmin` before we could check for `Inf` or `NaN`

Datseris replied:
Does it **error** or does it **warn**? In either case, that's unfortunate but oh well what can't we do. 

Feel free to devise a special version for `SimpleATsit` otherwise leave it `true`. 

At least `successful_step(integ)` should check what I said above, because this would work for e.g., discrete systems.

rusandris replied:
It gives an error and stops the execution. Indeed, `successful_step` should work that way, also for the discrete ones. Maybe I'll do that too

rusandris replied:
> At least successful_step(integ) should check what I said above, because this would work for e.g., discrete systems.

Added the checking (4th commit) and tested it with Stroboscopic map, Projected Integrator and discrete integrator. Since this is more general, I suggest moving this function from `continuous.jl`. Maybe.

Datseris commented on src/core/continuous.jl:
```diff
@@ -283,3 +283,10 @@ function SciMLBase.reinit!(integ::AbstractODEIntegrator, u0::AbstractVector,
+successful_step(integ::AbstractODEIntegrator{Alg, IIP, S, T}) where {Alg <:AbstractSimpleDiffEqODEAlgorithm, IIP, S, T} = true
+
+function successful_step(integ::AbstractODEIntegrator)
+	(integ.sol.retcode ==:Success || integ.sol.retcode ==:Default) ? true : false
```
```suggestion
	(integ.sol.retcode == :Success || integ.sol.retcode == :Default) ? true : false
```

Datseris commented on src/core/continuous.jl:
```diff
@@ -284,7 +284,10 @@ function SciMLBase.reinit!(integ::AbstractODEIntegrator, u0::AbstractVector,
 
-successful_step(integ) = true
+function successful_step(integ)
+	all(isfinite,integ.u) ? true : false
```
We need to use `any` here. If any entry of the vector is infinite we already have a problem. `all` doesn't cover this.

Also, please use proper API, i.e., `get_state(integ)` instead of `integ(u)`.

rusandris replied:
Nono, `all(isfinite,integ.u)` checks if **all** of them are **finite** , so we are on the same page. `any(isinfinite,integ.u)` would be the same. I'll correct the state part though.

Datseris replied:
Aaaaah sorry, my apologies, I missread your `isfinite` for `isinfinite`. Great :)

rusandris commented:
> Where did you test that and what exactly does this mean?

It just means I've tested those cases locally (checked which method is called and if the output is correct).

> you need to add tests to this pull request

Okay, that is what I was stuck on, so thanks for the help. I'll figure it out I hope.

Datseris commented:
great, thanks, I know it's annoying to always have to add something more in a "simple change", but I promise you it is the best for the longevity of the project!

rusandris commented:
I was putting together the testsets when I noticed that we have a somewhat bigger problem. Sometimes, `lyapunov` fails even when `successful_step` is there and it goes into an infinite loop silently. 
I was testing `lyapunov` with an unstable discrete system, and noticed that when the states blow up (but they're still finite, so `successful_step` still gives `true`), rescaling fails, `a` and `d` become NaNs and the function goes into an infinite loop because it never gets to `step!`   

Check this out:

```
function chaotictest!(du,u, p, t)
           k1, k2 = p
           x, y, z = u
           du[1] = y
           du[2] = z
           du[3] = k1-7*y+x^2+k2*z
       end
u0 = rand(3)
para = [2, -0.9] #k2=k6=0
dsd = DiscreteDynamicalSystem(chaotictest!, u0, para)
lyapunov(dsd,14)
```

Datseris commented:
Hm, interesting. I see. This does make sense though. The distance of the trajectories could be a larger number than the entries of the trajectories themselves. I guess we need to add one more checking statement that checks if `d` becomes `inf` or `NaN`, right? Then this would solve it I would imagine!

rusandris commented:
> I guess we need to add one more checking statement that checks if d becomes inf or NaN, right? Then this would solve it I would imagine!

I've added this change and also uploaded the tests. See the [ChaosTools pull request](https://github.com/JuliaDynamics/ChaosTools.jl/pull/262). Sorry, I accidentally made to commits instead of one.
I've also added here a testfile for the `succesful_step` alone, but I was unsure where to put it so I left it as a separate file. Let me know if something doesn't work.

rusandris commented:
It seems like we're getting the error
 `ERROR: LoadError: UndefVarError: AbstractSimpleDiffEqODEAlgorithm not defined`
I don't get that error locally using any dev packages of SimpleDiffEq (even if they're not modified at all), only when I use the globally installed one. So I'm absolutely lost here

rusandris commented:
Hmm it started working for me with SimpleDiffEq 1.6, no idea why. Can you trigger a check?

Datseris commented:
> Sorry, I accidentally made to commits instead of one.

No worries, we squash-merge everything anyways :) 

I'm a bit swamped with other projects unfortunately, but I promise you I'll return here and help you once we have I have a breath of fresh air!

Funkerfish commented:
Hi guys, really nice work! Have learned much from your codes, I wish someday I could contribute to this community.
I want to ask a question:
what you guys do here including the ```succesful_step``` is still detecting the unstable things, right? also like checking the ```NaN``` as @Datseris mentioned. But how can we interrupt the continuation once instability comes?

You know, for an unstable system with different initial conditions or parameter sets, number of steps to reach instability are quite different. Which means a large threshold is required for ```Step!``` to detect all unstable situations, for example ```Step!(integrator, 2000)```. I think it may also be responsible for @rusandris 's [```Lyapunov``` fails even when ```successful_step``` is there] situation.

When we use functions with large set of initial values and parameters, it's quite time-consuming to do all the checks, is there anything can help me terminate the continuation?

Should I post some code here?

rusandris commented:
> what you guys do here including the succesful_step is still detecting the unstable things, right? also like checking the NaN as @Datseris mentioned. But how can we interrupt the continuation once instability comes?

Well, in case of the AbstractODEIntegrators the `successful_step` instability checking relies on the internal `check_error!` function from `SciMLBase` (which tells us the return codes), in case of simple integrators there is just a placeholder function since these have an internal error throw for instability. For other integrators (for ex. discrete systems) we check for `NaNs` directly.

> Which means a large threshold is required for Step! to detect all unstable situations, for example Step!(integrator, 2000).

Note that we don't want to detect all unstable situations, we just want to know if a step fails inside one of our functions (for ex. in `lyapunov`) in which case we should break out immediately, which wasn't the case before. `step!(integ)` only makes one step, the size of which is determined internally by the solver (or externally if it's not adaptive).
I think `lyapunov` failed because the `d` distance between the two trajectories became so big it made the rescaling process to overflow.

> When we use functions with large set of initial values and parameters, it's quite time-consuming to do all the checks, is there anything can help me terminate the continuation?

I don't exactly understand what you mean here. You want to explore all stable parameter and initial condition regions for a system?

rusandris commented:
> I'm a bit swamped with other projects unfortunately, but I promise you I'll return here and help you once we have I have a breath of fresh air!

Thanks @Datseris ! You've really helped a lot, this was a first for me. Let me know when you're ready to finish this!

Datseris commented:
I'm sorry @rusandris surprisingly I'm currently more swamped than before.... But I've put this on my saved list and my TODOs, and once I return here I'll let you know.

Datseris commented:
Hello @rusandris I closed this to motivate opening a _new_ pull request that adds a function `successful_step(ds::DynamicalSystem) -> Bool` that returns `true` if the previous step was successful, false otherwise. It is by default `true` for all systems and you can add a dispatch for `CoupledODEs`. DOes this seem like something you could do?

Datseris commented:
(I just merged and tagged v3.0)
